### ISSUE TITLE
Process different size of text with CNN

### ISSUE THREAD

**Issue by McVilla** [fchollet/keras#3293]
Hi , I'm new to keras and I want to process different size of sentences for learning a document embedding in CNN. I used zero-padding and mask_out in Embedding layer, but layers after Embedding layer doesn't support masking. Is there any way to solve it?

`inputs = Input(shape=(1,max_len),dtype='int32')`
`x= Embedding(vocab_size, dim,weights = GloVe, mask_zero = True, input_length=max_len)(inputs)`
`x = Reshape((1,max_len,50))(x)`
`x = Convolution2D(nb_filter, n_gram, dim,init='glorot_uniform', activation='linear',border_mode='valid', subsample=(1,1))(x)`
`x = MaxPooling2D(pool_size=(2,1))(x1)`
`x = Flatten()(x)`
`out = Dense(10)(x)`
`Conv_sen= Model(inputs,out)`

Conv_sen is a shared model, I want to use it to precess every sentences in a document for learning their own sentence embedding. The following merge will concatenate them to form a document embedding.

`Inp_list = [None] * len(list_of_sentence)`
`for i in range(len(list_of_sentence)) :`
`Inp_list[j] = Input(shape=(1,max_len),dtype='int32')`
`sen_emb[j]= Conv_sen(Inp_list[j])`

`concat = merge(sen_emb, mode = 'concat')`

By the way, the max words in a sentence is about 250, and the average length is about 30. Is it ok to use zero padding for every sentence?
Thanks. 
Please make sure that the boxes below are checked before you submit your issue. Thank you!
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
  pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps
- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
  pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps
- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by mtreviso** [comment]
You can use bucket strategy with `train_on_batch` or `fit_generator`. It is necessary that your batches are homogeneous. Also, you have to set `max_len=None` in your model.

Pay attention on timestep size involved in conv layer. Depending the border_mode selected you may have some problems. 

Btw, you can use conv1d instead of conv2d :-). See this example: https://gist.github.com/mtreviso/7430af6d3c294c204dd2be1737ed284c

**Comment by McVilla** [comment]
Thanks, @mtreviso.  I can't do as the example, because the order of sentences matters.

**Comment by mtreviso** [comment]
@McVilla: if the order matters, then you can feed one sentence per batch using `train_on_batch`. It will be slow, but viable.

**Comment by iskandr** [comment]
This is a more fundamental problem with Keras's sequence handling. The successful embed->conv models I've seen seem to just treat the padding character as a usable vector value and set `mask_zero=False`. Not sure how much is lost by that hack.

**Comment by iskandr** [comment]
As far as I understand, this is still an open problem with Keras.
